Some clarifications about data persistence [HZG-257] (#1510)
hazelcast/hz-docs: docs/modules/storage/pages/persistence.adoc

= Persisting Data on Disk

> :description: Persistence allows individual members and whole clusters to recover data by persisting map entries, JCache data, and streaming job snapshots on disk. Members can use persisted data to recover from cluster-wide shutdowns and speed up individual member restarts by reducing the volume of data sent over the network.

Reviewer: I think your addition is helpful context to have, but I think the 3 scenarios mentioned are important and should not be replaced here. Your context relates to planned shutdowns/rolling upgrades, but sudden cluster-wide crashes and single member failures should be supported.

Author: So I kind of merged the planned and unplanned shutdowns/restarts together so I can separate those again. It's not clear that 

> Members can use persisted data to recover from ... a single member failure

is useful here. A cluster with backups enabled can already recover from a single member failure, persistence doesn't add that feature, it just speeds that process up for large volumes of data. Having this here makes it seem like persistence is an alternative to configuring in-memory backups.

Reviewer: Agreed that we want to be clear that persistence _can_ speed up single member failure recovery, but the backup mechanism is the primary recovery method 👍

> :description: Persistence allows individual members and whole clusters to recover data by persisting map entries, JCache data, and streaming job snapshots on disk. Members can use persisted data to recover from planned cluster-wide shutdowns, unplanned cluster-wide failures and speed up individual member restarts by reducing the volume of data sent over the network.

Reviewer: ```suggestion
:description: Persistence allows individual members and whole clusters to recover data by persisting map entries, JCache data, and streaming job snapshots on disk. Members can use persisted data to recover from planned cluster-wide shutdowns, unplanned cluster-wide failures, and to accelerate individual member restarts by reducing the volume of data sent over the network.
```
minor rewording for clarity

:toc-levels: 3
:page-enterprise: true

{description}

== Why Persist Data

> Data in Hazelcast is usually stored in-memory (RAM) so that it's faster to access. However, data in RAM is volatile, meaning that when a member shuts down, its data is lost. In-memory partition backups are still the primary method for data resilience in Hazelcast but additionally persisting data to disk can be useful in certain situations. When you persist data on disk, members can load it upon a restart back into memory directly instead of relying on other active members sending it over the network.

Reviewer: ```suggestion
Data in Hazelcast is usually stored in-memory (RAM) so that it's faster to access. However, data in RAM is volatile, meaning that when a member shuts down, its data is lost. In-memory partition backups are the primary method for data resilience in Hazelcast but additionally persisting data to disk can be useful in certain situations. When you persist data on disk, members can load it back into memory upon a restart  instead of relying on other active members sending it over the network.
```

Some wording tweaks - since this is under "why persist data", it's probably worth taking this opportunity to add a bit of context about the benefit of persistence in the event of a cluster-wide failure (i.e. DC loses power), as it allows recovery of data that would otherwise be irreversibly lost.

Clusters can use persisted data for the following scenarios:

> - **Persisting data across cluster-wide shutdowns**:

Reviewer: ```suggestion
- **Cluster-wide shutdowns**:
```
Nit: we already start this list with "clusters can use persisted data for...", so I don't think we need to have "Persisting data across" here

> ** **Planned**: A whole cluster is shut down and restarted with the same configuration, state, and data as before the restart. 

Reviewer: I'm not sure I agree with the removal of information here - when the cluster state is `PASSIVE`, it is perfectly fine to perform a rolling restart with persistence.

Author: I didn't remove it, I moved it under the single member restart category. Since a rolling restart is a sequence of single member restarts instead of the cluster restarting.

+
[IMPORTANT]
====
In order for persistence to function correctly for planned cluster shutdowns you must xref:maintain-cluster:shutdown.adoc[shutdown the entire cluster at once].

> ====

Reviewer: This is not accurate, the cluster state must be `PASSIVE` and then it is acceptable to shut down members individually (MC will move the cluster state to `PASSIVE` for restarts, for example).

Author: Yeah this could probably be worded better, the link I included is about how to shutdown an entire cluster by putting it into passive mode first and then shutting down the members, one of the ways in that doc is using MC. I can reword it to be more explicit about the passive state and then provide the link as a list of ways you might go about shutting down the cluster.

Author: Reworded this, let me know what you think.

> ** **Unplanned**: A cluster is restarted after all its members crash at the same time due an event such as a power outage. Note that some data loss is expected unless `fsync` is set to `true`. For more information, see xref:storage:configuring-persistence.adoc#data-structures[Data structure options].

Reviewer 2: ```suggestion
** **Unplanned**: A cluster is restarted after all its members crash at the same time due to an event such as a power outage. Note that some data loss is expected unless `fsync` is set to `true`. For more information, see xref:storage:configuring-persistence.adoc#data-structures[Data structure options].
```

- **Speeding up single member restarts**: 

> ** **Planned**: During a rolling restart each cluster member is restarted one by one for scenarios such as installing an operating system patch or new hardware. xref:maintain-cluster:rolling-upgrades.adoc[Rolling upgrades] are an example of a rolling restart.

Reviewer: This is part of a cluster-wide shutdown, we do not want to create confusion between these topics imo

Author: I found it confusing to read the existing way, how is a rolling restart a cluster wide shutdown if only one member at any time is stopped? Is it mandatory to put the cluster into passive mode for a rolling restart? It seems like it work fine if the cluster was still in active mode.

Reviewer: I suppose there's arguments to be made both ways - I can see why you see it as a series of single member restarts, but I can also see that it does _eventually_ result in the entire cluster restarting, so it could be seen as a cluster-wide shutdown (restart). I'm not too fussed, we can let docs team review this aspect and see if it makes sense to them 👍

> - **Accelerating single member restarts**: 

Reviewer 2: ```suggestion
- **Single member restarts**: 
```

The two bullets should be consistent. Removing the verb here; could alternatively add a verb ("recovering from cluster-wide shutdowns") above.

> - **Speeding up single member restarts**: 

Reviewer: ```suggestion
- **Accelerating single member restarts**: 
```
Nit: more appealing wording, at least imo

> ** **Unplanned**: A member may crash or terminate unexpectedly at any time, using persistence allows faster recovery.

Reviewer: This also depends on configuration - if rebalancing is delayed, or the cluster is in `PASSIVE` state, it loads all data from disk without receiving any from other members

Author: I was under the impression that even if the member crashed when the cluster was in active state and rebalancing was not delayed the persisted data could still be used to speed up the member rejoining the cluster. Aren't merkle trees compared to reduce network traffic as much as possible?

Reviewer: Yes that's correct, I'm saying that in addition to this "faster recovery" case, there are conditions where it recovers entirely using its own data, not just using it as a speedup mechanism.

> ** **Unplanned**: A member may crash or terminate unexpectedly at any time, using persistence allows faster recovery.

Reviewer: ```suggestion
** **Unplanned**: A single member may crash or terminate unexpectedly at any time, using persistence allows faster recovery.
```
We only support single member failure in terms of persistence functionality
